### ISSUE TITLE
1.22 regression: removing and recreating static pod manifest leaves pod in error state

### ISSUE THREAD

**Issue by liggitt** [kubernetes/kubernetes#104648]
#### What happened:

* Started a 1.22 cluster
* Added a static pod manifest
* Observed the pod run successfully
* Removed and recreated the pod manifest
* Observed the pod enter an error state and never restart successfully

#### What you expected to happen:

For the pod to be re-run successfully (as in previous releases)

#### How to reproduce it (as minimally and precisely as possible):

```
# start a cluster
hack/local-up-cluster.sh

# create a static pod
echo '
kind: Pod
apiVersion: v1
metadata:
  name: mypod
spec:
  terminationGracePeriodSeconds: 1
  containers:
  - name: busybox
    image: busybox
    command: ["sh", "-c", "echo $RANDOM; sleep 100000"]
' > mypod.yaml

sudo cp mypod.yaml /var/run/kubernetes/static-pods/

# wait until the pod is running
watch kubectl get pods -n default

# remove and recreate the static pod
sudo rm /var/run/kubernetes/static-pods/mypod.yaml
sudo cp mypod.yaml /var/run/kubernetes/static-pods/

# observe the pod remain in an error state
watch kubectl get pods -n default
```

#### Anything else we need to know?:

Bisected to 3eadd1a9ead7a009a9abfbd603a5efd0560473cc (https://github.com/kubernetes/kubernetes/pull/102344)

Broken in 1.22, works in 1.21 and previous versions.

/sig node
cc @smarterclayton @bobbypage @rphillips

**Comment by liggitt** [comment]
I also verified the pending PR fixing the pod admission regression in https://github.com/kubernetes/kubernetes/issues/104560 (https://github.com/kubernetes/kubernetes/pull/104611) does not fix this issue

**Comment by liggitt** [comment]
kubelet log

<details>

```
I0829 01:50:33.287717 2909862 kubelet.go:2045] "SyncLoop REMOVE" source="file" pods=[default/mypod-127.0.0.1]
I0829 01:50:33.287842 2909862 kubelet.go:1890] "Pod has been deleted and must be killed" pod="default/mypod-127.0.0.1" podUID=7c0663d52f9e8c61c7a733304c23b60f
I0829 01:50:33.287957 2909862 kubelet_pods.go:1363] "Generating pod status" pod="default/mypod-127.0.0.1"
I0829 01:50:33.288248 2909862 status_manager.go:472] "Ignoring same status for pod" pod="default/mypod-127.0.0.1" status={Phase:Running Conditions:[{Type:Initialized Status:True LastProbeTime:0001-01-01 00:00:00 +0000 UTC LastTransitionTime:2021-08-29 01:48:34 +0000 UTC Reason: Message:} {Type:Ready Status:True LastProbeTime:0001-01-01 00:00:00 +0000 UTC LastTransitionTime:2021-08-29 01:48:37 +0000 UTC Reason: Message:} {Type:ContainersReady Status:True LastProbeTime:0001-01-01 00:00:00 +0000 UTC LastTransitionTime:2021-08-29 01:48:37 +0000 UTC Reason: Message:} {Type:PodScheduled Status:True LastProbeTime:0001-01-01 00:00:00 +0000 UTC LastTransitionTime:2021-08-29 01:48:34 +0000 UTC Reason: Message:}] Message: Reason: NominatedNodeName: HostIP:127.0.0.1 PodIP:192.168.9.3 PodIPs:[{IP:192.168.9.3}] StartTime:2021-08-29 01:48:34 +0000 UTC InitContainerStatuses:[] ContainerStatuses:[{Name:busybox State:{Waiting:nil Running:&ContainerStateRunning{StartedAt:2021-08-29 01:48:36 +0000 UTC,} Terminated:nil} LastTerminationState:{Waiting:nil Running:nil Terminated:nil} Ready:true RestartCount:0 Image:busybox:latest ImageID:docker-pullable://busybox@sha256:b37dd066f59a4961024cf4bed74cae5e68ac26b48807292bd12198afa3ecb778 ContainerID:docker://868ff9744cd94c3ae1fc041f227152e31e18028aefb85a06b924c2c4b533c780 Started:0xc0013e007a}] QOSClass:BestEffort EphemeralContainerStatuses:[]}
I0829 01:50:33.288437 2909862 kuberuntime_container.go:712] "Killing container with a grace period override" pod="default/mypod-127.0.0.1" podUID=7c0663d52f9e8c61c7a733304c23b60f containerName="busybox" containerID="docker://868ff9744cd94c3ae1fc041f227152e31e18028aefb85a06b924c2c4b533c780" gracePeriod=1
I0829 01:50:33.288482 2909862 kuberuntime_container.go:716] "Killing container with a grace period" pod="default/mypod-127.0.0.1" podUID=7c0663d52f9e8c61c7a733304c23b60f containerName="busybox" containerID="docker://868ff9744cd94c3ae1fc041f227152e31e18028aefb85a06b924c2c4b533c780" gracePeriod=1
I0829 01:50:33.288586 2909862 event.go:291] "Event occurred" object="default/mypod-127.0.0.1" kind="Pod" apiVersion="v1" type="Normal" reason="Killing" message="Stopping container busybox"
I0829 01:50:33.333676 2909862 file.go:201] "Reading config file" path="/var/run/kubernetes/static-pods/mypod.yaml"
I0829 01:50:33.334496 2909862 file.go:201] "Reading config file" path="/var/run/kubernetes/static-pods/mypod.yaml"
I0829 01:50:33.334638 2909862 kubelet.go:2035] "SyncLoop ADD" source="file" pods=[default/mypod-127.0.0.1]
I0829 01:50:33.947082 2909862 scope.go:111] "RemoveContainer" containerID="53f8c8f1c0a5c987efe1657e589be108a90f10cc7f25fca41cd7f9338c0747ca"
I0829 01:50:33.955351 2909862 kubelet_pods.go:1030] "Clean up pod workers for terminated pods"
I0829 01:50:33.955415 2909862 kubelet_pods.go:1056] "Clean up probes for terminating and terminated pods"
I0829 01:50:33.959053 2909862 kubelet_pods.go:1093] "Clean up orphaned pod statuses"
I0829 01:50:33.961802 2909862 kubelet_pods.go:1112] "Clean up orphaned pod directories"
I0829 01:50:33.961938 2909862 kubelet_pods.go:1123] "Clean up orphaned mirror pods"
I0829 01:50:33.961970 2909862 kubelet_pods.go:1130] "Clean up orphaned pod cgroups"
I0829 01:50:34.381431 2909862 manager.go:1044] Destroyed container: "/kubepods/besteffort/pod7c0663d52f9e8c61c7a733304c23b60f/868ff9744cd94c3ae1fc041f227152e31e18028aefb85a06b924c2c4b533c780" (aliases: [k8s_busybox_mypod-127.0.0.1_default_7c0663d52f9e8c61c7a733304c23b60f_0 868ff9744cd94c3ae1fc041f227152e31e18028aefb85a06b924c2c4b533c780], namespace: "docker")
I0829 01:50:34.504559 2909862 kuberuntime_container.go:724] "Container exited normally" pod="default/mypod-127.0.0.1" podUID=7c0663d52f9e8c61c7a733304c23b60f containerName="busybox" containerID="docker://868ff9744cd94c3ae1fc041f227152e31e18028aefb85a06b924c2c4b533c780"
I0829 01:50:34.506376 2909862 plugins.go:433] "Calling network plugin to tear down the pod" pod="default/mypod-127.0.0.1" networkPluginName="kubernetes.io/no-op"
I0829 01:50:34.559772 2909862 manager.go:1044] Destroyed container: "/kubepods/besteffort/pod7c0663d52f9e8c61c7a733304c23b60f/143f0e3bb4dadbdc224e518719d475acc7c3fe2792105323ea9f5627b786d6ee" (aliases: [k8s_POD_mypod-127.0.0.1_default_7c0663d52f9e8c61c7a733304c23b60f_0 143f0e3bb4dadbdc224e518719d475acc7c3fe2792105323ea9f5627b786d6ee], namespace: "docker")
I0829 01:50:35.469173 2909862 generic.go:296] "Generic (PLEG): container finished" podID=7c0663d52f9e8c61c7a733304c23b60f containerID="868ff9744cd94c3ae1fc041f227152e31e18028aefb85a06b924c2c4b533c780" exitCode=137
I0829 01:50:35.469236 2909862 kubelet.go:2073] "SyncLoop (PLEG): event for pod" pod="default/mypod-127.0.0.1" event=&{ID:7c0663d52f9e8c61c7a733304c23b60f Type:ContainerDied Data:868ff9744cd94c3ae1fc041f227152e31e18028aefb85a06b924c2c4b533c780}
I0829 01:50:35.469188 2909862 kubelet_pods.go:1363] "Generating pod status" pod="default/mypod-127.0.0.1"
I0829 01:50:35.469294 2909862 kubelet.go:2073] "SyncLoop (PLEG): event for pod" pod="default/mypod-127.0.0.1" event=&{ID:7c0663d52f9e8c61c7a733304c23b60f Type:ContainerDied Data:143f0e3bb4dadbdc224e518719d475acc7c3fe2792105323ea9f5627b786d6ee}
I0829 01:50:35.469317 2909862 pod_container_deletor.go:79] "Container not found in pod's containers" containerID="143f0e3bb4dadbdc224e518719d475acc7c3fe2792105323ea9f5627b786d6ee"
I0829 01:50:35.469402 2909862 volume_manager.go:446] "Waiting for volumes to unmount for pod" pod="default/mypod-127.0.0.1"
I0829 01:50:35.469444 2909862 volume_manager.go:473] "All volumes are unmounted for pod" pod="default/mypod-127.0.0.1"
I0829 01:50:35.474651 2909862 manager.go:1044] Destroyed container: "/kubepods/besteffort/pod7c0663d52f9e8c61c7a733304c23b60f" (aliases: [], namespace: "")
I0829 01:50:35.490279 2909862 status_manager.go:621] "Patch status for pod" pod="default/mypod-127.0.0.1" patch="{\"metadata\":{\"uid\":\"42fa471d-43c1-46ce-af40-196062d2efb5\"},\"status\":{\"$setElementOrder/conditions\":[{\"type\":\"Initialized\"},{\"type\":\"Ready\"},{\"type\":\"ContainersReady\"},{\"type\":\"PodScheduled\"}],\"conditions\":[{\"lastTransitionTime\":\"2021-08-29T01:50:35Z\",\"message\":\"containers with unready status: [busybox]\",\"reason\":\"ContainersNotReady\",\"status\":\"False\",\"type\":\"Ready\"},{\"lastTransitionTime\":\"2021-08-29T01:50:35Z\",\"message\":\"containers with unready status: [busybox]\",\"reason\":\"ContainersNotReady\",\"status\":\"False\",\"type\":\"ContainersReady\"}],\"containerStatuses\":[{\"containerID\":\"docker://868ff9744cd94c3ae1fc041f227152e31e18028aefb85a06b924c2c4b533c780\",\"image\":\"busybox:latest\",\"imageID\":\"docker-pullable://busybox@sha256:b37dd066f59a4961024cf4bed74cae5e68ac26b48807292bd12198afa3ecb778\",\"lastState\":{},\"name\":\"busybox\",\"ready\":false,\"restartCount\":0,\"started\":false,\"state\":{\"terminated\":{\"containerID\":\"docker://868ff9744cd94c3ae1fc041f227152e31e18028aefb85a06b924c2c4b533c780\",\"exitCode\":137,\"finishedAt\":\"2021-08-29T01:50:34Z\",\"reason\":\"Error\",\"startedAt\":\"2021-08-29T01:48:36Z\"}}}]}}"
I0829 01:50:35.491913 2909862 status_manager.go:630] "Status for pod updated successfully" pod="default/mypod-127.0.0.1" statusVersion=3 status={Phase:Running Conditions:[{Type:Initialized Status:True LastProbeTime:0001-01-01 00:00:00 +0000 UTC LastTransitionTime:2021-08-29 01:48:34 +0000 UTC Reason: Message:} {Type:Ready Status:False LastProbeTime:0001-01-01 00:00:00 +0000 UTC LastTransitionTime:2021-08-29 01:50:35 +0000 UTC Reason:ContainersNotReady Message:containers with unready status: [busybox]} {Type:ContainersReady Status:False LastProbeTime:0001-01-01 00:00:00 +0000 UTC LastTransitionTime:2021-08-29 01:50:35 +0000 UTC Reason:ContainersNotReady Message:containers with unready status: [busybox]} {Type:PodScheduled Status:True LastProbeTime:0001-01-01 00:00:00 +0000 UTC LastTransitionTime:2021-08-29 01:48:34 +0000 UTC Reason: Message:}] Message: Reason: NominatedNodeName: HostIP:127.0.0.1 PodIP:192.168.9.3 PodIPs:[{IP:192.168.9.3}] StartTime:2021-08-29 01:48:34 +0000 UTC InitContainerStatuses:[] ContainerStatuses:[{Name:busybox State:{Waiting:nil Running:nil Terminated:&ContainerStateTerminated{ExitCode:137,Signal:0,Reason:Error,Message:,StartedAt:2021-08-29 01:48:36 +0000 UTC,FinishedAt:2021-08-29 01:50:34 +0000 UTC,ContainerID:docker://868ff9744cd94c3ae1fc041f227152e31e18028aefb85a06b924c2c4b533c780,}} LastTerminationState:{Waiting:nil Running:nil Terminated:nil} Ready:false RestartCount:0 Image:busybox:latest ImageID:docker-pullable://busybox@sha256:b37dd066f59a4961024cf4bed74cae5e68ac26b48807292bd12198afa3ecb778 ContainerID:docker://868ff9744cd94c3ae1fc041f227152e31e18028aefb85a06b924c2c4b533c780 Started:0xc00165e9ea}] QOSClass:BestEffort EphemeralContainerStatuses:[]}
I0829 01:50:35.495845 2909862 status_manager.go:621] "Patch status for pod" pod="default/mypod-127.0.0.1" patch="{\"metadata\":{\"uid\":\"42fa471d-43c1-46ce-af40-196062d2efb5\"}}"
I0829 01:50:35.495928 2909862 status_manager.go:628] "Status for pod is up-to-date" pod="default/mypod-127.0.0.1" statusVersion=4
```

</details>


kubelet's view of the pod (/pods kubelet endpoint)

<details>

```json
{
  "kind": "PodList",
  "apiVersion": "v1",
  "metadata": {},
  "items": [
    {
      "metadata": {
        "name": "mypod-127.0.0.1",
        "namespace": "default",
        "selfLink": "/api/v1/namespaces/default/pods/mypod-127.0.0.1",
        "uid": "7c0663d52f9e8c61c7a733304c23b60f",
        "creationTimestamp": null,
        "annotations": {
          "kubernetes.io/config.hash": "7c0663d52f9e8c61c7a733304c23b60f",
          "kubernetes.io/config.seen": "2021-08-29T01:50:33.334534656Z",
          "kubernetes.io/config.source": "file"
        }
      },
      "spec": {
        "containers": [
          {
            "name": "busybox",
            "image": "busybox",
            "command": [
              "sh",
              "-c",
              "echo $RANDOM; sleep 100000"
            ],
            "resources": {},
            "terminationMessagePath": "/dev/termination-log",
            "terminationMessagePolicy": "File",
            "imagePullPolicy": "Always"
          }
        ],
        "restartPolicy": "Always",
        "terminationGracePeriodSeconds": 1,
        "dnsPolicy": "ClusterFirst",
        "nodeName": "127.0.0.1",
        "securityContext": {},
        "schedulerName": "default-scheduler",
        "tolerations": [
          {
            "operator": "Exists",
            "effect": "NoExecute"
          }
        ],
        "enableServiceLinks": true
      },
      "status": {
        "phase": "Running",
        "conditions": [
          {
            "type": "Initialized",
            "status": "True",
            "lastProbeTime": null,
            "lastTransitionTime": "2021-08-29T01:48:34Z"
          },
          {
            "type": "Ready",
            "status": "False",
            "lastProbeTime": null,
            "lastTransitionTime": "2021-08-29T01:50:35Z",
            "reason": "ContainersNotReady",
            "message": "containers with unready status: [busybox]"
          },
          {
            "type": "ContainersReady",
            "status": "False",
            "lastProbeTime": null,
            "lastTransitionTime": "2021-08-29T01:50:35Z",
            "reason": "ContainersNotReady",
            "message": "containers with unready status: [busybox]"
          },
          {
            "type": "PodScheduled",
            "status": "True",
            "lastProbeTime": null,
            "lastTransitionTime": "2021-08-29T01:48:34Z"
          }
        ],
        "hostIP": "127.0.0.1",
        "podIP": "192.168.9.3",
        "podIPs": [
          {
            "ip": "192.168.9.3"
          }
        ],
        "startTime": "2021-08-29T01:48:34Z",
        "containerStatuses": [
          {
            "name": "busybox",
            "state": {
              "terminated": {
                "exitCode": 137,
                "reason": "Error",
                "startedAt": "2021-08-29T01:48:36Z",
                "finishedAt": "2021-08-29T01:50:34Z",
                "containerID": "docker://868ff9744cd94c3ae1fc041f227152e31e18028aefb85a06b924c2c4b533c780"
              }
            },
            "lastState": {},
            "ready": false,
            "restartCount": 0,
            "image": "busybox:latest",
            "imageID": "docker-pullable://busybox@sha256:b37dd066f59a4961024cf4bed74cae5e68ac26b48807292bd12198afa3ecb778",
            "containerID": "docker://868ff9744cd94c3ae1fc041f227152e31e18028aefb85a06b924c2c4b533c780",
            "started": false
          }
        ],
        "qosClass": "BestEffort"
      }
    }
  ]
}
```

</details>

**Comment by liggitt** [comment]
this seems likely to be caused by the pod cleanup being keyed by uid and not handling an overlapping delete operation and recreate operation for a pod with the same uid

the calculated uid for static pods consists of:
* node name
* file name
* pod content

https://github.com/kubernetes/kubernetes/blob/edb0a72cff0e43bab72a02cada8486d562ee1cd5/pkg/kubelet/config/common.go#L57-L70

Removing and then replacing the manifest resulted in a pod with an identical computed uid.

Changing any of the manifest inputs to the pod uid calculation resulted in the second pod starting successfully:
* removing the first manifest and creating an identical manifest with a different filename
* removing the first manifest and creating a manifest with the same name with a slightly different pod (e.g. a random annotation inserted)

Also, waiting for the pod deletion to complete for a few seconds, then readding the manifest resulting in an identical pod UID worked.

**Comment by liggitt** [comment]
I can imagine schemes for including a generation var for static manifest files, where every time the kubelet observes a deletion of a manifest file, it bumps the generation of that file, and generations above 0 are included in the computed pod uid.

Is it important for the kubelet's calculated uid for a static pod to remain stable over kubelet restarts (e.g. does it get baked into a persistent attribute in the CRI and compared by the kubelet in a way that a kubelet restart that changed the computed uid would restart the container)?

**Comment by jayunit100** [comment]
maybe another hostPRocess test for this to generate this scenario ?
/sig windows

**Comment by rphillips** [comment]
I tested this with latest master [d1da6d47dd9dc8eeee9b96bffa9523e4c91f32b8] and cannot reproduce this issue. Can anyone else?

**Comment by rphillips** [comment]
I tested https://github.com/kubernetes/kubernetes/pull/104031 and it does not fix this issue.

**Comment by gjkim42** [comment]
How about including file modification time when calculating the static pod's uid.
It would prevent this issue, and calculated uid would be stable over kubelet restarts.

However, we still need to solve #97722 too.

**Comment by rphillips** [comment]
@gjkim42 That is a good idea. I do worry about some controllers not being aware of the modified time which would cause restarts of the static pods, but I've PRed the patch to test it in CI: https://github.com/kubernetes/kubernetes/pull/104712.

**Comment by liggitt** [comment]
@deads2k, thoughts on the modified time approach?

note that this won't help for static pods which explicitly set a uid

**Comment by gjkim42** [comment]
https://kubernetes.slack.com/archives/C0BP8PW9G/p1625072555009600

I remember there was a discussion about whether we would support explicitly set uid in static pods.
And maybe we decided to not support that.

---

One concern is that if we calculate a uid for a static pod including file modification time,
including pod content(previous approach) may be meaningless.
We may remove the property that kubelet does not re-create static pods with the same content.

**Comment by neolit123** [comment]
> How about including file modification time when calculating the static pod's uid.
It would prevent this issue, and calculated uid would be stable over kubelet restarts.

when moving a file away from a directory and moving it back in after some time (which is effectively a static pod restart) or when renaming it, the modification time would remain the same. thus modification time would not contribute to the uniqueness of the UUID. this is consistent on Linux and Windows.

"change time" or  `st_ctime` of the POSIX `stat` can be considered, but it is a NO-OP on Windows as there is no "change time" concept there.

another example about OSes differences is is that the kubelet periodic re-opening a file and checking contents means "access time" (`st_atime`) should be changing on Linux. but on Windows/ NTFS the property is disabled by default for performance reasons, meaning the system admin has to first enable it. both "access time" and "change time" are missing in Go's `os.FileInfo` because they are not portable concepts.

**Comment by rphillips** [comment]
@neolit123 Thank you for that information. I changed the PR to mix in a timestamp, but this isn't going to work either.

**Comment by smarterclayton** [comment]
Note that we need exclusion by name/namespace for static pods AND regular pods in force deletion - we shouldn't start the new pod with the same name and different UID when there is an old one terminating (and force deleted).  So changing the UID to be generated isn't sufficient, but it's additional.

**Comment by smarterclayton** [comment]
Thinking through this problem, I think the fundamental gap is that desired config is not being properly translated into "desired running" and proper exclusion isn't being provided.  I *think* that there should be an exclusion rule which is:

1. two pods with same uid cannot be running at the same time (pod config enforces this today except for force deleted pods)
2. two pods with the same source (file / api) and the same full name cannot be running at the same time

We would also need to enforce

3. a static pod that changes names without uid changing must be treated as being deleted and recreated by pod config (so the pod worker knows that the old pod must start going away and this isn't an update)

If we fix those, a static pod would be able to ensure either exclusion or inclusion by using pod full name to exclude duplicates, and uid changes to force exclusion.  That would also address the bug in question here (the race that the pod worker can lose the pod because the pod worker has to deal with pods that are force deleted AND have duplicate UIDs).

**Comment by gjkim42** [comment]
> The guarantee UID for static pods _was_ providing was "don't allow two instances of this static pod with the same UID to be running at the same time", which autogenerating UID would result in being allowed.

Right, we can't solve this bug by generating UID.



> 3. a static pod that changes names without uid changing must be treated as being deleted and recreated by pod config (so the pod worker knows that the old pod must start going away and this isn't an update)

I think there can't be a static pod that could change the name without uid changing. Am I misunderstood?

**Comment by gjkim42** [comment]
@smarterclayton 

> 2\. two pods with the same source (file / api) and the same full name cannot be running at the same time

If kubelet does not allow two pods from apiserver with the same full name, I think force deleting statefulset pods might not work as before. I am not sure it is ok.

**Comment by ehashman** [comment]
/remove-priority important-soon
/priority critical-urgent

Rationale: this is a critical regression in the lifecycle of static pods in 1.22. Once fixed, this should be backported as soon as we determine the patch is stable.

**Comment by 249043822** [comment]
Before v1.22, while a static pod is removed, it's podworker will be removed right now, although it's worker go-rountine may has not finished. just while the same static pod manifest comes back, the new podworker could be created, there may a slot time， old and new podworker coexist.
After v1.22, rescontructed the podworker, while the old static pod has not terminated completely, the come backed new static pod will be ignored

**Comment by 249043822** [comment]
@smarterclayton @liggitt @ehashman 

Can we fix this by making the new static pod start to create after the old one with same uid  terminated completely?  and the pseudo code may like below:

1. dispatchWork check the static pod create and waits for notification
```
// dispatchWork starts the asynchronous sync of the pod in a pod worker.
// If the pod has completed termination, dispatchWork will perform no action.
func (kl *Kubelet) dispatchWork(pod *v1.Pod, syncType kubetypes.SyncPodType, mirrorPod *v1.Pod, start time.Time) {
	if syncType == kubetypes.SyncPodCreate && kubetypes.IsStaticPod(pod) {
		if kl.podWorkers.HasStaticPodWaiting(pod.UID) {
			return
		}
		if kl.podWorkers.IsStaticPodNeedWaiting(pod.UID) {
			ch := make(chan struct{}, 1)
			kl.podWorkers.SetStaticPodWaiting(pod.UID, ch)
			go func() {
				select {
				case <-ch:
					// Run the sync in an async worker.
					kl.podWorkers.UpdatePod(UpdatePodOptions{
						Pod:        pod,
						MirrorPod:  nil,
						UpdateType: kubetypes.SyncPodCreate,
						StartTime:  start,
					})
					// Note the number of containers for new pods.
					metrics.ContainersPerPodCount.Observe(float64(len(pod.Spec.Containers)))
				}
			}()
			return
		}
	}
	// Run the sync in an async worker.
	kl.podWorkers.UpdatePod(UpdatePodOptions{
		Pod:        pod,
		MirrorPod:  mirrorPod,
		UpdateType: syncType,
		StartTime:  start,
	})
	// Note the number of containers for new pods.
	if syncType == kubetypes.SyncPodCreate {
		metrics.ContainersPerPodCount.Observe(float64(len(pod.Spec.Containers)))
	}
}
```
2. removeTerminatedWorker close the ch to notify new static pod creating
```

func (p *podWorkers) removeTerminatedWorker(uid types.UID) {
	status, ok := p.podSyncStatuses[uid]
	if !ok {
		// already forgotten, or forgotten too early
		klog.V(4).InfoS("Pod worker has been requested for removal but is not a known pod", "podUID", uid)
		return
	}

	if !status.finished {
		klog.V(4).InfoS("Pod worker has been requested for removal but is still not fully terminated", "podUID", uid)
		return
	}

	klog.V(4).InfoS("Pod has been terminated and is no longer known to the kubelet, remove all history", "podUID", uid)
	delete(p.podSyncStatuses, uid)
	delete(p.podUpdates, uid)
	delete(p.lastUndeliveredWorkUpdate, uid)
	if ch, ok := p.waitingStaticPods[uid]; ok {
		close(ch)
	}
	delete(p.waitingStaticPods, uid)

}

func (p *podWorkers) SetStaticPodWaiting(uid types.UID, ch chan<- struct{}) {
	p.podLock.Lock()
	defer p.podLock.Unlock()
	p.waitingStaticPods[uid] = ch
}

func (p *podWorkers) IsStaticPodNeedWaiting(uid types.UID) bool {
	p.podLock.Lock()
	defer p.podLock.Unlock()
	if status, ok := p.podSyncStatuses[uid]; ok {
		return status.IsTerminated() || status.IsTerminationRequested() || status.IsTerminationStarted()
	}
	return false
}

func (p *podWorkers) HasStaticPodWaiting(uid types.UID) bool {
	p.podLock.Lock()
	defer p.podLock.Unlock()
	if _, ok := p.waitingStaticPods[uid]; ok {
		return true
	}
	return false
}
```

if this thought can be approved, I can do this work

**Comment by smarterclayton** [comment]
The rough structure you described is similar to #104847, except we need to do this for non static pods as well.  There are a couple of problems related to a change like this though because the rest of the Kubelet ONLY references a pod by UID.  So if you were to rapidly delete a pod and create a new one with the same UID AND completely change volumes or devices, it's very likely that other parts of the Kubelet could perform some of their cleanup or tear down logic inconsistently.  I *think* volumes are correct (although that's only if setup and teardown are happening in serial), but I haven't reviewed the others.  Just waiting some unit of time isn't enough - we need to either guarantee those loops are safe for two separate pod definitions with the same UID to be passed to them in quick succession (better testing + review) or we need to add a construct that ensures that those loops don't think the pod is the same.

The real tension here is that reusing a UID for a completely different instance of a pod (like static pods allow) is not desirable, because UID is supposed to represent an "instance of a pod" and pods provide certain one way guarantees around lifecycle (once a pod is failed, it stays that way forever).  The behavior of static pods shouldn't be modelled like a pod that lives forever, it should be modelled using a controller that stamps out new desired pods (each with their own UID).  However, that's a behavior change, and it's of course possible (but incredibly unlikely) for a duplicate UID to be generated on the apiserver, so the easiest path probably is just "making static pods work".

In which case, the design of what we're trying to do in Kubelet is:

1. Ensure that the pod worker runs at most one  pod definition with a given uid at a time
2. Ensure that when a pod with a given UID exits, if the pod is still desired with that UID, the new one is started in the pod worker

That's hard to do outside the pod worker today because the loop that drives pod worker is edge driven (change to pod config), not level driven.  We'd need to fix the loop from pod config to pod worker to reconcile the desired state against the actual state, and the kubelet doesn't directly track desired state (because admission decisions aren't recorded in a single list).  But we can't do it from within the pod worker, because we have no idea when the pod worker receives the sequence `STOP(uid=1) -> CREATE(uid=1) -> STOP(uid=1)` whether the second stop cancels the `CREATE` or is simply shortening the first `STOP`.

Probably having pod housekeeping filter the list of desired pods and then have it start any pod that is still desired but is terminated in the Kubelet is the right fix, but I need to think through the tradeoffs some more.

**Comment by smarterclayton** [comment]
https://docs.google.com/document/d/1NJKYNgoXZKGS5la4MvGrB42-DNQSFf4JQkEcZRnrDzA/edit# is the doc that captures all of the implications here.

**Comment by smarterclayton** [comment]
https://github.com/kubernetes/kubernetes/pull/104847 is now updated with what I believe is a more complete fix for this issue, but I'm still testing the assumptions in the Kubelet.

**Comment by ehashman** [comment]
> this seems likely to be caused by the pod cleanup being keyed by uid and not handling an overlapping delete operation and recreate operation for a pod with the same uid
> 
> Removing and then replacing the manifest resulted in a pod with an identical computed uid.

Note that identical UIDs on static pods pose a problem for calculating pod lifecycle of static pods, tracked in https://bugzilla.redhat.com/show_bug.cgi?id=1933760#c27 

Why the static pod UIDs were not changing between boots was a source of confusion for me---if pods are mortal, surely the static pod should get a new UID after dying/terminating, I reasoned, regardless of the unchanged name/manifest.

> The guarantee UID for static pods was providing was "don't allow two instances of this static pod with the same UID to be running at the same time", which autogenerating UID would result in being allowed.

Aha, I see.

> The real tension here is that reusing a UID for a completely different instance of a pod (like static pods allow) is not desirable, because UID is supposed to represent an "instance of a pod" and pods provide certain one way guarantees around lifecycle (once a pod is failed, it stays that way forever).

This matches my intuition above.

I think we should try to find a way to ensure:

- when a static pod is terminated, a new instance of that pod doesn't reuse the same UID
- we still maintain the guarantee that two static pods with identical name/namespace cannot run simultaneously

I think the UID reuse here is ultimately the root of the bug, and it breaks assumptions not just here but in the rest of the kubelet.

**Comment by ehashman** [comment]
> Note that identical UIDs on static pods pose a problem for calculating pod lifecycle of static pods, tracked in https://bugzilla.redhat.com/show_bug.cgi?id=1933760#c27
>
> I think the UID reuse here is ultimately the root of the bug, and it breaks assumptions not just here but in the rest of the kubelet.

Elaborating on this a little bit... Right now, we see the following behaviour when a node reboots:

- Static pods terminate, node shuts down
- Node starts up, kubelet does its reconciliation loop
- Static pods are launched prior to API server sync and enter pending state, if the container runtime does not retain information on the previous static pod status [true on OpenShift because we call cri-o wipe, but this could also happen on manual wipe or during disaster recovery]
- New static pods use the same UID as the manifests are unchanged
- Kubelet eventually syncs with API server. Mirror pod with the same UID used to be running, but if it doesn't launch fast enough, kubelet will sync a new Pending status
- Pod illegally transitions back from Running->Pending

So IMO the reuse of the UID is a bit of a hack that breaks core assumptions of the kubelet.

**Comment by smarterclayton** [comment]
> Pod illegally transitions back from Running->Pending

Turns out this actually is not an illegal transition (which I realized during dealing with soft admission).  A pod that fails soft admission post start is put back into pending because it's not actually running and the kubelet has decided to hold it.  It's strictly better for the kubelet to communicate this because the alternative is a user has no idea that the pod wouldn't be running.  So I think going forward Running -> Pending is a bug when  we're losing state (i.e. restart with wipe) or when we know there is a concrete transition (soft admitted to soft not-admitted)
